### PR TITLE
Support TTL for index

### PR DESCRIPTION
**Pull request by bright-starry-sky**
Not ready to review

### CODE REVIEW

**Comment by bright-starry-sky** [comment]
TODO : 

1. test cases for storage side.
2. correct test cases for graph side.

**Comment by bright-starry-sky** [comment]
Added test cases for storage side.

**Comment by bright-starry-sky** [comment]
test cases changed for graph side : 
https://github.com/vesoft-inc/nebula-graph/pull/876

**Comment by bright-starry-sky** [comment]
Ready to review.

**Review by critical27 — COMMENTED** [review]
LGTM, good job. Any related PR?

**Comment by bright-starry-sky** [comment]
> LGTM, good job. Any related PR?

After this one merged , test case need to correct for graph side : https://github.com/vesoft-inc/nebula-storage/pull/382

File: src/storage/CommonUtils.cpp [hunk 1]
```diff
@@ -109,5 +114,31 @@ bool CommonUtils::checkDataExpiredForTTL(const meta::SchemaProviderIf* schema,
...
+    std::string col;
+    if (sp.get_ttl_col()) {
+        col = *sp.get_ttl_col();
+    }
+    return std::make_pair(!(duration <= 0 || col.empty()), std::make_pair(duration, col));
+}
+
+StatusOr<Value> CommonUtils::ttlValue(const meta::SchemaProviderIf* schema, RowReader* reader) {
+    DCHECK(schema != nullptr);
+    const auto* ns = dynamic_cast<const meta::NebulaSchemaProvider*>(schema);
```
Comment (critical27) [inline src/storage/CommonUtils.cpp:135]: We could remove `SchemaProviderIf` later if necessary, it is useless now.

**Review by critical27 — APPROVED** [review]

File: src/storage/CommonUtils.cpp (hunk 1 above)
Comment (bright-starry-sky) [reply src/storage/CommonUtils.cpp:135]: > We could remove `SchemaProviderIf` later if necessary, it is useless now.

Yes, I see other place that need to be changed, Let me change them later.

File: src/storage/admin/RebuildTagIndexTask.cpp [hunk 2]
```diff
@@ -97,6 +97,31 @@ kvstore::ResultCode RebuildTagIndexTask::buildIndexGlobal(GraphSpaceID space,
...
         }
 
+        auto schema = env_->schemaMan_->getTagSchema(space, tagID);
+        if (!schema) {
+            LOG(WARNING) << "Space " << space << ", tag " << tagID << " invalid";
+            iter->next();
+            continue;
+        }
+
+        auto ttlProp = CommonUtils::ttlProps(schema.get());
```
Comment (darionyaphet) [inline src/storage/admin/RebuildTagIndexTask.cpp:107]: Could  move to rebuild index task ？

Comment (bright-starry-sky) [reply src/storage/admin/RebuildTagIndexTask.cpp:107]: > Could move to rebuild index task ？

What's the purpose?

**Comment by bright-starry-sky** [comment]
Fixed compile error and rebased master.

**Review by critical27 — APPROVED** [review]

File: src/meta/processors/schemaMan/AlterTagProcessor.cpp [hunk 3]
```diff
@@ -63,6 +63,23 @@ void AlterTagProcessor::process(const cpp2::AlterTagReq& req) {
...
+    if (existIndex) {
+        int64_t duration = 0;
+        if (prop.get_ttl_duration()) {
+            duration = *prop.get_ttl_duration();
+        }
+        std::string col;
+        if (prop.get_ttl_col()) {
+            col = *prop.get_ttl_col();
+        }
+        if (!col.empty() && duration > 0) {
```
Comment (panda-sheep) [inline src/meta/processors/schemaMan/AlterTagProcessor.cpp:75]: Can ttl and index coexist?

Comment (bright-starry-sky) [reply src/meta/processors/schemaMan/AlterTagProcessor.cpp:75]: Aha, hey can coexist. I saved the timestamp value to index val.

Comment (panda-sheep) [reply src/meta/processors/schemaMan/AlterTagProcessor.cpp:75]: `existIndex `  means there is an index,
`!col.empty() && duration> 0` means there is ttl

Why is there an error here?

Comment (bright-starry-sky) [reply src/meta/processors/schemaMan/AlterTagProcessor.cpp:75]: > `existIndex ` means there is an index,
> `!col.empty() && duration> 0` means there is ttl
> 
> Why is there an error here?

if index exists,  not allow to alter TTL properties.

Comment (panda-sheep) [reply src/meta/processors/schemaMan/AlterTagProcessor.cpp:75]: yeah, i know what you mean.
But the `prop ` schema prop is in the `schema`, not the schema prop in `AlterTagReq`.

Your current logic is that when alter tag, as long as there is index and ttl, an error will be reported.

**Review by panda-sheep — COMMENTED** [review]
Good job.
could you add a UT that a tag has already a index and ttl, and then alter this tag（not alter ttl prop）？

File: src/meta/processors/schemaMan/AlterTagProcessor.cpp (hunk 3 above)
Comment (bright-starry-sky) [reply src/meta/processors/schemaMan/AlterTagProcessor.cpp:75]: > yeah, i know what you mean.
> But the `prop ` schema prop is in the `schema`, not the schema prop in `AlterTagReq`.
> 
> Your current logic is that when alter tag, as long as there is index and ttl, an error will be reported.

yes，there should be an error thrown here, either `alter ttl ` or `alter column` (excepted for `add column` )

**Comment by bright-starry-sky** [comment]
> Good job.
> could you add a UT that a tag has already a index and ttl, and then alter this tag（not alter ttl prop）？

Aha, the old test cases are still valid for storage side. for graph side, the below PR  to do with what you said.

**Comment by bright-starry-sky** [comment]
> > Good job.
> > could you add a UT that a tag has already a index and ttl, and then alter this tag（not alter ttl prop）？
> 
> Aha, the old test cases are still valid for storage side. for graph side, the below PR to do with what you said.

vesoft-inc/nebula-graph#876

File: src/meta/processors/schemaMan/AlterTagProcessor.cpp (hunk 3 above)
Comment (panda-sheep) [reply src/meta/processors/schemaMan/AlterTagProcessor.cpp:75]: I mean:
here logic is unreasonable.
As long as there are index and ttl,  here are not allowed to alter tag

Comment (bright-starry-sky) [reply src/meta/processors/schemaMan/AlterTagProcessor.cpp:75]: > I mean:
> here logic is unreasonable.
> As long as there are index and ttl, here are not allowed to alter tag

good point. 🐂B
artimg

**Comment by bright-starry-sky** [comment]
Addressed panda-sheep's comments.

**Review by panda-sheep — APPROVED** [review]
Good job！

**Review by critical27 — APPROVED** [review]
Good job!

**Comment by bright-starry-sky** [comment]
This file `src/meta/test/ProcessorTest.cpp` is too large and often causes the link to fail on CI system (ubuntu1804, gcc-9.2).
I'm going to split it in the next PR.

**Pull request merged**